keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -0,0 +1,401 @@
+
+A GRPC based interface will be used by CSI Node Plugins to invoke privileged operations through csi-proxy.exe. The GRPC API will be versioned and any release of csi-proxy.exe binary will maintain backward compatibility across all prior versions of the API. This avoids having to run multiple versions of the csi-proxy.exe binary on the same Windows host if multiple CSI Node Plugins have been configured and the version of the csi-proxy API required by the plugins are wide apart.
+
+The following are the main RPC calls that will comprise a v1 version of the csi-proxy API:
+
+```
+service CSIProxyService {
+
+    // Version returns the earliest and latest version of the CSI Proxy API supported by the service.
+    rpc Version(VersionRequest) returns (VersionResponse) {}
+
+    // Mkdir creates a directory at the requested absolute path in the host. Relative path is not supported.
+    rpc Mkdir(MkdirRequest) returns (MkdirResponse) {}
+
+    // Rmdir removes the directory at the requested absolute path in the host. Relative path is not supported.
+    // This may be used for unlinking a symlink created through LinkVolume
+    rpc Rmdir(RmdirRequest) returns (RmdirResponse) {}
+
+    // StageDisk partitions a disk device and formats the partition with NTFS (if the disk has not been partitioned or formatted)
+    // The resulting volume is mounted at the requested global staging path.
+    rpc StageDisk(StageDiskRequest) returns (StageDiskResponse) {}
+
+    // StageSMBShare mounts a remote share over SMB on the host at the requested global staging path.
+    rpc StageSMBShare(StageSMBShareRequest) returns (StageSMBShareResponse) {}
+
+    // LinkVolume invokes mklink on the global staging path of a volume linking it to a path within a container
+    rpc LinkVolume(LinkVolumeRequest) returns (LinkVolumeResponse) {}
+
+    // GetDiskNumberWithLocation finds a disk device with requested Location <Bus, Target, LUN ID>
+    rpc GetDiskNumberWithLocation(GetDiskNumberWithLocationRequest) returns (GetDiskNumberWithLocationResponse) {}
+
+    // GetDiskNumberWithSCSIID finds a disk device whose SCSI Page 83 ID contains the passed ID string
+    rpc GetDiskNumberWithSCSIID(GetDiskNumberWithSCSIIDRequest) returns (GetDiskNumberWithSCSIIDResponse) {}
+}
+```
+
+In accordance with standard Kubernetes conventions, the above API will be introduces as v1alpha1 and graduate to v1beta1 and v1 as the feature graduates. New RPCs and enhancements to parameters will be introduced through v1alpha1 and graduate to v1beta1 and v1 versions as they mature.
+
+#### Enhancements in CSI Node Plugins
+
+Code for CSI Node Plugins need to be refactored to support CSI Node APIs in both Linux and Windows nodes. While the code targeting Linux nodes can assume privileged access to the host, the code targeting Windows nodes need to invoke the GRPC client API associated with the desired version of the CSIProxyService described above.
```

Comment:
Will the same CSI plugin work for both Linux and Windows, or will separate codebase/images be maintained?

Reply:
Images will definitely be separate as the Windows plugin will be based on Windows base images and the Linux node plugin will be based on a Linux distro of choice. The codebase can be same. I will point this out specifically.